### PR TITLE
[NFC] Switch a number of DenseMaps to SmallDenseMaps for speedup

### PR DESCRIPTION
**Pull request by jmorse**
tl;dr, if we use SmallDenseMaps instead of DenseMaps at these locations, we get a substantial speedup because there's less spurious malloc traffic:

https://llvm-compile-time-tracker.com/compare.php?from=983635c014767a2a37f87a4b564b59a78b866154&to=1198cb4aa9c1715d18212a21328801b6985b8ca7&stat=instructions:u

Background: inspired by @SLTozer 's introspective collection of stacktraces for some debug-info things, I've instrumented DenseMap to print where it was allocated and the max number of elements it contained. Run over CTMark and with the addition of some filtering, this has picked out the locations in LLVM where we allocate a DenseMap hashtable off the heap but we could instead get away with using the inline buckets of a SmallDenseMap and avoid calling malloc. I picked 16 inline elements at callsites which occasionally have more than 12 elements inserted, and four inline elements for some callsites where there typically aren't any elements inserted.

One drawback of this technique is that it's fully tuned to making the compile-time-tracker happy, so might not be representative in general. Counterpoints would be that CTMark is chosen to have a range of different inputs and is vaguely representative, avoiding allocations is almost always a win, and in scenarios where we will /always/ insert at least one element it makes sense to spend a little stack memory to avoid that.

(I've got two more patches that contribute another ~0.3% speedup, but it's now hit diminishing returns).

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/ScheduleDAGFast.cpp [hunk 1]
```diff
@@ -770,7 +770,7 @@ void ScheduleDAGLinearize::Schedule() {
 MachineBasicBlock*
 ScheduleDAGLinearize::EmitSchedule(MachineBasicBlock::iterator &InsertPos) {
   InstrEmitter Emitter(DAG->getTarget(), BB, InsertPos);
-  DenseMap<SDValue, Register> VRBaseMap;
+  SmallDenseMap<SDValue, Register, 16> VRBaseMap;
```
Comment (nikic) [inline llvm/lib/CodeGen/SelectionDAG/ScheduleDAGFast.cpp:773]: Base `VRBaseMapType` public and use `InstrEmitter::VRBaseMapType` here?

**Review by nikic — APPROVED** [review]
LGTM. As long as it's not inside nested maps, there's basically no downside to having inline storage.

Also nice results on clang, so this clearly does generalize beyond CTMark:
```
bin/clang-20 	6913606M 	6858510M (-0.80%)
bin/llvm-tblgen 	261829M 	261017M (-0.31%)
bin/clang-tblgen 	121228M 	118689M (-2.09%)
```
